Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -348,14 +355,204 @@ lemma isIndContained_iff_exists_iso_subgraph :
+    G.copyCount H = 1 := by
+  rw [copyCount]
+  convert card_singleton (⊥ : H.Subgraph)
+  simp only [eq_singleton_iff_unique_mem, mem_filter, mem_univ, Subgraph.coe_bot, true_and,
+    Nonempty.forall, Subsingleton.elim G ⊥]
+  haveI : IsEmpty (⊥ : H.Subgraph).verts := by simp
+  refine ⟨⟨⟨⟨isEmptyElim, isEmptyElim, isEmptyElim, isEmptyElim⟩, fun {a} ↦ isEmptyElim a⟩⟩,
+    fun H' e ↦ Subgraph.ext ?_ <| funext₂ fun a b ↦ ?_⟩
+  · simpa [Set.eq_empty_iff_forall_not_mem, filter_eq_empty_iff, ‹IsEmpty V›] using
+      e.toEquiv.symm.isEmpty_congr
+  · simp only [Subgraph.not_bot_adj, eq_iff_iff, iff_false]
+    exact fun hab ↦ e.symm.map_rel_iff.2 hab.coe
+
+@[simp] lemma copyCount_eq_zero : G.copyCount H = 0 ↔ ¬ G ⊑ H := by
+  simp [copyCount, -nonempty_subtype, isContained_iff_exists_iso_subgraph, card_pos,
+    filter_eq_empty_iff]
+
+@[simp] lemma copyCount_pos : 0 < G.copyCount H ↔ G ⊑ H := by
+  simp [copyCount, -nonempty_subtype, isContained_iff_exists_iso_subgraph, card_pos,
+    filter_nonempty_iff]
+
+end CopyCount
+
+section LabelledCopyCount
+variable [Fintype V] [Fintype W]
+
+/-- `G.labelledCopyCount H` is the number of labelled copies of `G` in `H`. See
+`SimpleGraph.copyCount` for the number of unlabelled copies. -/
+noncomputable def labelledCopyCount (G : SimpleGraph V) (H : SimpleGraph W) : ℕ := by
+  classical exact Fintype.card (Copy G H)
+
+@[simp] lemma labelledCopyCount_of_isEmpty [IsEmpty V] (G : SimpleGraph V) (H : SimpleGraph W) :
+    G.labelledCopyCount H = 1 := by
+  convert Fintype.card_unique
+  exact { default := ⟨default, isEmptyElim⟩, uniq := fun _ ↦ Subsingleton.elim _ _ }
+
+@[simp] lemma labelledCopyCount_eq_zero : G.labelledCopyCount H = 0 ↔ ¬ G ⊑ H := by
+  simp [labelledCopyCount, IsContained, Fintype.card_eq_zero_iff, isEmpty_subtype]
+
+@[simp] lemma labelledCopyCount_pos : 0 < G.labelledCopyCount H ↔ G ⊑ H := by
+  simp [labelledCopyCount, IsContained, Fintype.card_pos_iff]
+
+/-- There's more labelled copies of `H` of-`G` than unlabelled ones. -/
+lemma copyCount_le_labelledCopyCount : G.copyCount H ≤ G.labelledCopyCount H := by
+  classical
+  rw [copyCount, ← Fintype.card_coe]
+  refine Fintype.card_le_of_surjective (fun f ↦ ⟨f.toSubgraph, ?_⟩) ?_
+  · simpa using ⟨f.isoToSubgraph⟩
+  · simpa [Surjective, -Copy.range_toSubgraph] using fun H' e ↦ Copy.toSubgraph_surjOn ⟨e⟩
+
+end LabelledCopyCount
+
+/-!
+#### Induced copies
+
+TODO
 
-@[simp] lemma compl_isIndContained_compl : Gᶜ ⊴ Hᶜ ↔ G ⊴ H :=
-  Embedding.complEquiv.symm.nonempty_congr
+### Killing a subgraph
 
-protected alias ⟨IsIndContained.of_compl, IsIndContained.compl⟩ := compl_isIndContained_compl
+An important aspect of graph containment is that we can remove not too many edges from a graph `H`
+to get a graph `H'` that doesn't contain `G`.
+
+#### Killing not necessarily induced copies
+
+`SimpleGraph.kill G H` is a subgraph of `H` where an edge was removed from each copy of `G` in `H`.
```

Comment:
I find the argument order here quite confusing, Usually `G.operation` is a modification of `G`, whereas here it's a modification of the other argument using `G`. What do you think?

Reply:
I don't hugely mind, but I guess it is true that `G.kill H` reads more like "`G`, where you kill `H`" than vice-versa. Furthermore, `H` is often used as the graph you're trying to avoid in the literature